feat(runtime): Lua ftplugin 'includeexpr'

brianhuster opened the pull request:
Problem: Current `'includeexpr'` in runtime/ftplugin/lua.vim doesn't work with Nvim Lua.

Solution: Better includeexpr for Lua ftplugin

Closes: https://github.com/neovim/neovim/issues/32490

TODO:
- [x] Use `vim.loader.find()`
- [x] Update news.txt
- [x] Add tests 
~- [ ] Refer to `vim.lua_includeexpr` in `:h ft-lua`~

justinmk commented:
https://github.com/neovim/neovim/issues/32490#issuecomment-2676981414 why can't this live in the lua ftplugin ?

clason commented:
If need be, move it to a `runtime/lua/_ftplugin.lua` (exact name is placeholder, except for the mandatory underscore; if we expect more "autoload at home", maybe `/_ftplugin/lua.lua`).

There may be justification for the "autoload" pattern ("only require what you use, when you use it"), but indeed there is zero reason for polluting the `vim` namespace with these functions.

brianhuster commented:
I want the function to be public, so that users can set `'includeexpr'` to it.

So maybe I should move it to `vim/ftplugin.lua` or `vim/lua.lua` file?

brianhuster commented:
You mean I should set it in `runtime/ftplugin/lua.lua`? I didn't think of that, because I think users may use Lua for other purposes than Nvim config and plugins.

Or maybe I should add a global variable to allow users to switch between Nvim Lua and normal Lua?

clason commented:
Then the function should be made smarter to be a strict improvement on the current Vimscript `includexpr`. Having to switch would be a dealbreaker here.

brianhuster commented:
Another question. Should I also add a function to find root or should I just assume that the current directory is the root of the Lua project?

zeertzjq commented on runtime/lua/vim/_editor.lua:
```diff
@@ -1142,6 +1142,7 @@ do
   end
 end
 
+
```
```suggestion
```

justinmk commented on runtime/ftplugin/lua.lua:
```diff
@@ -1,4 +1,8 @@
 vim.treesitter.start()
 
-vim.b.undo_ftplugin = (vim.b.undo_ftplugin or '') .. '\n call v:lua.vim.treesitter.stop()'
+vim.bo.includeexpr = "v:lua.require'_ftplugin.lua'.includeexpr(v:fname)"
```
Is `_ftplugin.lua` needed because `runtime/ftplugin/` is not on 'runtimepath' ? 

Rather than introducing `_ftplugin.lua`, can we use `loadfile(vim.env.VIMRUNTIME..'/ftplugin/lua.lua')` ?  This should work:

    vim.bo.includeexpr = "v:lua.loadfile($VIMRUNTIME..'/ftplugin/lua.lua')"

clason replied:
No, it's just an arbitrary path to avoid polluting the global namespace (and the top-level `runtime/lua` directory). It's not about _this_ `ftplugin` but about the "autoload" file for the includeexpr (which needs to live in a different file for this to work without headaches).

You can replace `require` with `loadfile`, but this will give you no benefit (and cost you `require`'s caching).

zeertzjq commented on .gitignore:
```diff
@@ -84,3 +84,5 @@ tags
 /contrib/result-*
 
 CMakeUserPresets.json
+
+Xconfirm_write_ro
```
No

brianhuster committed:
feat(runtime): Lua ftplugin sets 'omnifunc', 'foldexpr' #32697

Problem:
- Many other ftplugin have defined 'omnifunc', but the Lua one doesn't
  define one, even though there is `vim.lua_omnifunc()`
- Users may want "stupid" completion to fix Lua config with
  `nvim --clean` in case they breaks it
- Nvim doesn't port Lua foldexpr from Vim

Solution:
- Set 'omnifunc' to 'v:lua.vim.lua_omnifunc' in ftplugin/lua.lua
- Set 'foldexpr' to use treesitter

justinmk commented on test/functional/lua/ftplugin_spec.lua:
```diff
@@ -0,0 +1,105 @@
+    )
+  end)
+
+  it('must returns path to module in packpath/start', function()
```
other cases too

```suggestion
  it('returns path to module in packpath/start', function()
```

justinmk commented on test/functional/lua/ftplugin_spec.lua:
```diff
@@ -0,0 +1,105 @@
+    )
+
+    n.command [[e `=$HOME . '/.nvim/lua/runtime-foo/init.lua'`]]
+    n.command('w ++p')
```
usually helpful for tests to use non-abbreviated command/option names.

also, could use a single, multiline `n.command[[...]]` for readability, for many parts of these tests?

```suggestion
    n.command('write ++p')
```

justinmk commented on test/functional/lua/ftplugin_spec.lua:
```diff
@@ -0,0 +1,105 @@
+  return exec_lua [[return vim.fs.normalize(vim.env.VIMRUNTIME)]]
+end
+
+describe('lua.includeexpr', function()
```
option names are usually single-quoted by convention (or `&foo`)

```suggestion
describe("Lua 'includeexpr'", function()
```

justinmk commented on test/functional/lua/ftplugin_spec.lua:
```diff
@@ -0,0 +1,105 @@
+local t = require('test.testutil')
```
this test file should probably live in `test/functional/editor/`. 

`lua/` tests are for the Lua stdlib. Not general features. The fact that this feature is implemented via `_ftplugin.lua` is an implementation detail.

justinmk reviewed (approved):
LGTM after comments.

justinmk commented on runtime/lua/_ftplugin/lua.lua:
Just noticed something else: to avoid conflicts with e.g. a plugin (or user config) that happens to be named `_ftplugin`, we should namespace any builtin Lua modules under `vim`.

Thus this file should live at `runtime/lua/vim/_ftplugin/lua.lua`.

(Sorry if this (possibly) conflicts with a previous suggestion that I made, though hopefully not.)

(Note: the top-level `editorconfig.lua`, `man.lua`, `tohtml.lua` plugins are "true plugins", that's why they have claimed their respective "namespaces". However, I wonder if we should reconsider that, maybe they should have a `nvim` or `vim` prefix. There's no real need for use to claim more and more top-level namespaces...)

brianhuster replied:
> However, I wonder if we should reconsider that, maybe they should have a nvim or vim prefix.

I always think so. Users may want to use another implementation of the plugin.

Vim has a few plugins that use top-level namespace like `comment`, `editorconfig`, `helptoc`, but they are all optional plugins that must be "installed" explicitly with `packadd`, so they less likely conflict.

justinmk commented on runtime/doc/news.txt:
```diff
@@ -251,6 +251,10 @@ DEFAULTS
   • Lua |ftplugin| sets 'omnifunc' to "v:lua.vim.lua_omnifunc".
   • Lua |ftplugin| sets 'foldexpr' to "v:lua.vim.treesitter.foldexpr()".
 
+• Options:
```
```suggestion
```

justinmk commented on runtime/doc/news.txt:
```diff
@@ -251,6 +251,10 @@ DEFAULTS
   • Lua |ftplugin| sets 'omnifunc' to "v:lua.vim.lua_omnifunc".
   • Lua |ftplugin| sets 'foldexpr' to "v:lua.vim.treesitter.foldexpr()".
 
+• Options:
+  • Lua |ftplugin| sets 'omnifunc' to "v:lua.vim.lua_omnifunc".
+  • Lua |ftplugin| sets 'foldexpr' to "v:lua.vim.treesitter.foldexpr()".
```
looks like a duplicate.

also, there are now 3 "Options" sections. (look just above)

justinmk commented on test/functional/editor/include_spec.lua:
the old name `ftplugin_spec.lua` was more discoverable. `include_spec.lua` isn't anything that will be recognized when looking for test categories.

if `fplugin_spec.lua` grows into a very big file that should later be split into more-granular concerns, that's a problem we will solve when we come to it (if ever). until then, hyper-categorizing is counterproductive.

justinmk commented on test/functional/editor/include_spec.lua:
```diff
@@ -0,0 +1,109 @@
+end
+
+describe("Lua 'includeexpr'", function()
+  it('returns path to module in the project', function()
```
```suggestion
  it('finds module in the project', function()
```

justinmk commented on test/functional/editor/include_spec.lua:
```diff
@@ -0,0 +1,109 @@
+    )
+  end)
+
+  it('returns path to module in VIMRUNTIME', function()
```
```suggestion
  it('finds module in VIMRUNTIME', function()
```

justinmk commented on test/functional/editor/ftplugin_spec.lua:
```diff
@@ -0,0 +1,109 @@
+local eq = t.eq
+local root = n.exec_lua('return vim.fs.normalize(vim.fn.getcwd())')
+
+before_each(function()
+  n.command [[
+    edit `=stdpath('config') .. '/lua/user-foo/init.lua'`
```
this will only work if `clear()` was previously called in some other test. 

while we're here, I suggest moving this into the describe() block. that will be needed in the future anyway, if/when other ftplugin-related tests are added to this file.

```suggestion
setup(n.clear)
before_each(function()
  n.command [[
    edit `=stdpath('config') .. '/lua/user-foo/init.lua'`
```

brianhuster commented:
Can you please help me resolve the conflict? I think I don't know Git enough

brianhuster committed:
feat(ftplugin): includeexpr for Nvim Lua

Problem: Current `'includeexpr'` in runtime/ftplugin/lua.vim doesn't work with Nvim Lua.

Solution: Better includeexpr for Lua ftplugin

Closes: https://github.com/neovim/neovim/issues/32490

Co-authored-by: Justin M. Keyes <justinkz@gmail.com>

justinmk committed:
Merge branch 'master' into vim.lua_includeexpr

justinmk commented:
resolved conflict, but lint might fail because `runtime/ftplugin/lua.lua` does not end with a newline 😭  (because of github's online text editor...)

can fix this by doing `:write` on that file in nvim, then commit and push.

justinmk commented on test/functional/editor/ftplugin_spec.lua:
```diff
@@ -0,0 +1,115 @@
+  setup(n.clear)
+  before_each(function()
+    command [[
+      edit `=stdpath('config') .. '/lua/user-foo/init.lua'`
```
nit: usually `exe 'edit' ..` is the preferred way to do this. `=` expressions are somewhat deprecated.

also typically `fnameescape()` is used to avoid path quirks.

brianhuster replied:
Hm, that's sad for me, I actually hope Nvim would support those backtick expansion more, because IMO they are shorter so better for using in cmdline.

Recently Helix also add support expansing cmdline https://github.com/helix-editor/helix/pull/12527. I notice `sh%{}`, which is already covered by Vim's backtick expansion

justinmk replied:
backtick-expansion won't be removed, it's just not common (and there are some quirks but I don't remember the details).

Note that inline Lua expressions are planned: https://github.com/neovim/neovim/issues/11306

brianhuster replied:
So only `backtick=` is deprecated?

justinmk replied:
Even in vim they are rarely used. It's just guidance. They aren't going to be removed.

brianhuster replied:
I rewrite those lines like this
```vim
      exe "edit" stdpath('config') .. '/lua/user-foo/init.lua'
      write ++p
      exe "edit" stdpath('config') .. '/lua/user-foo/bar.lua'
      write ++p
      exe "edit" stdpath('data') .. '/site/pack/packer/start/plugin-foo/lua/plugin-foo/init.lua'
      write ++p
      exe "edit" stdpath('data') .. '/site/pack/packer/start/plugin-foo/lua/plugin-foo/bar.lua'
      write ++p
```
But the test gives error `Undefined variable config`. so I revert to using `backtick=`

justinmk commented on test/functional/editor/ftplugin_spec.lua:
```diff
@@ -0,0 +1,115 @@
+      edit runtime/lua/foo/bar/init.lua
+      write ++p
+
+      edit ~/.nvim/lua/runtime-foo/init.lua
+      write ++p
+      edit ~/.nvim/lua/runtime-foo/bar.lua
```
any reason not to put this in `test/functional/fixtures/` ?

or use `t.tmpname(true)`. generally not a good idea to mess with `~/` in tests.

brianhuster commented:
Btw, should I squash the merge commit?

justinmk commented on test/functional/editor/ftplugin_spec.lua:
```diff
@@ -0,0 +1,85 @@
+    ]]):format(temp_dir, temp_dir))
+  end)
+
+  it('find module in current repo', function()
```
```suggestion
  it('finds module in current repo', function()
```

justinmk commented:
Thanks for driving this to a good conclusion. Will be useful :)

justinmk commented on runtime/ftplugin/lua.lua:
```diff
@@ -1,9 +1,10 @@
 -- use treesitter over syntax
 vim.treesitter.start()
 
+vim.bo.includeexpr = 'v:lua.require"vim._ftplugin.lua".includeexpr()'
```
For future reference: if/when options support function refs, then we could avoid the need for `vim._ftplugin`, like this:

```lua
vim.bo.includeexpr = function()
  ...
end
```

justinmk commented:
followup https://github.com/neovim/neovim/pull/32948
